### PR TITLE
Feature: add rack-attack to protect sign-in from brute force

### PR DESCRIPTION
**Pull request by Kohelbekker**

### CODE REVIEW

**Comment by chumaknadya** [comment]
> ⚠️ We detected 3 security issues in this pull request:
> 
> Vulnerable Libraries (3)
> 
> Severity	Details
> N/A	[actionpack@5.2.4.4](https://github.com/openware/barong/blob/11fa8a09fb48ad5b82c08fd0d3f26e0f6ab3cd72/Gemfile.lock#L13) upgrade to `~> 5.2.4.6, ~> 5.2.6, ~> 6.0.3.7, >= 6.1.3.2`
> High	[nokogiri@1.10.10](https://github.com/openware/barong/blob/11fa8a09fb48ad5b82c08fd0d3f26e0f6ab3cd72/Gemfile.lock#L56) upgrade to `>= 1.11.4`
> High	[puma@3.12.6](https://github.com/openware/barong/blob/11fa8a09fb48ad5b82c08fd0d3f26e0f6ab3cd72/Gemfile.lock#L317) upgrade to `~> 4.3.8, >= 5.3.1`
> More info on how to fix Vulnerable Libraries in [Ruby](https://docs.guardrails.io/docs/en/vulnerabilities/ruby/using_vulnerable_libraries.html?utm_source=ghpr).
> 
> 👉 Go to the [dashboard](https://dashboard.guardrails.io/gh/openware/30086/scans?sha=11fa8a09fb48ad5b82c08fd0d3f26e0f6ab3cd72) for detailed results.
> 
> 📥 Happy? Share your [feedback](https://guardrails.typeform.com/to/BCgBkB) with us.

@Kohelbekker Could you fix these issues, by upgrading a gems to those version

File: config/initializers/rack_attack.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    throttle('logins/email', limit: 5, period: 60.seconds) do |req|
+      if req.path.include?('/identity/sessions') && req.post?
+        # Normalize the email, using the same logic as your authentication process, to
+        # protect against rate limit bypasses. Return the normalized email if present, nil otherwise.
+        params = JSON.parse( req.body.read )
+        params['email'].to_s.downcase.gsub(/\s+/, "").presence
+      end
+    end
+
+    ### Custom Throttle Response ###
```
Comment (chumaknadya) [inline config/initializers/rack_attack.rb:65]: Lets create a doc about new changes, so everyone could follow update

File: config/initializers/rack_attack.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    # attack where an attacker simply tries a large number of emails and
+    # passwords to see if any credentials match.
+    #
+    # Another common method of attack is to use a swarm of computers with
+    # different IPs to try brute-forcing a password for a specific account.
+
+    # Throttle POST requests to /login by IP address
+    #
+    # Key: "rack::attack:#{Time.now.to_i/:period}:logins/ip:#{req.ip}"
+    throttle('logins/ip', limit: 5, period: 60.seconds) do |req|
```
Comment (chumaknadya) [inline config/initializers/rack_attack.rb]: lets make env variable for limit cout

**Review by chumaknadya — CHANGES_REQUESTED** [review]

File: docs/security.md [hunk 3]
```diff
@@ -53,3 +53,13 @@ For further information check the links down below:
...
 * https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#javascript-guidance-for-auto-inclusion-of-csrf-tokens-as-an-ajax-request-header
 * https://www.owasp.org/index.php/Category:OWASP_Top_Ten_Project
+
+### Sign in brute force protection
+
+To protect those users who don't have OTP enabled from brute-force login attacks we use `rack-attack` gem.
+
+By default Barong have protection of `identity/sessions` endpoint, which throttle POST requests to this endpoint by both IP address and email. Limit of requests per 60 seconds can be configured by changing of `barong_rack_attack_limit` env (default value is 5). If limit exceeded then user will receive `429` repsonse and will be able to successfully retry when 60 seconds will have passed from the first attempt.
+
+In `test` or `development` environment rack-attack protection is dissabled unless `ENABLE_RACK_ATTACK` env is set to `true`.
```
Comment (chumaknadya) [inline docs/security.md]: disabled

File: docs/security.md [hunk 4]
```diff
@@ -53,3 +53,13 @@ For further information check the links down below:
 * https://www.owasp.org/index.php/Main_Page
 * https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#javascript-guidance-for-auto-inclusion-of-csrf-tokens-as-an-ajax-request-header
 * https://www.owasp.org/index.php/Category:OWASP_Top_Ten_Project
+
+### Sign in brute force protection
+
+To protect those users who don't have OTP enabled from brute-force login attacks we use `rack-attack` gem.
```
Comment (chumaknadya) [inline docs/security.md:59]: for now we have this check even for users with otp, as we are using same endpoint?

File: config/initializers/rack_attack.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
+class Rack::Attack
+
+    Rack::Attack.enabled = Rails.env.production? || ENV.fetch('ENABLE_RACK_ATTACK', false)
```
Comment (chumaknadya) [inline config/initializers/rack_attack.rb]: could we prefix this env BARONG

**Review by chumaknadya — CHANGES_REQUESTED** [review]

File: docs/security.md (hunk 4 above)
Comment (Kohelbekker) [reply docs/security.md:59]: Yes, we have it for all users, I've added this part just to make it clear that users with otp protected against brute force by otp itself.

**Review by chumaknadya — APPROVED** [review]

**Pull request merged**